diff -uNr tor-0.2.3.15-alpha/src/or/config.c tor-0.2.3.15-alpha-persistent_keys_in_etc/src/or/config.c
--- tor-0.2.3.15-alpha/src/or/config.c	2012-04-29 17:07:09.000000000 -0400
+++ tor-0.2.3.15-alpha-persistent_keys_in_etc/src/or/config.c	2012-05-21 13:05:10.332574460 -0400
@@ -1458,17 +1458,13 @@
 
   /* If needed, generate a new TLS DH prime according to the current torrc. */
   if (server_mode(options) && options->DynamicDHGroups) {
-    char *keydir = get_datadir_fname("keys");
-    if (check_private_dir(keydir, CPD_CREATE, options->User)) {
-      tor_free(keydir);
+    char *keydir = "/etc/tor/var";
+    if (check_private_dir(keydir, CPD_CREATE, options->User))
       return -1;
-    }
-    tor_free(keydir);
 
     if (!old_options || !old_options->DynamicDHGroups) {
-      char *fname = get_datadir_fname2("keys", "dynamic_dh_params");
+      char *fname = "/etc/tor/var/dynamic_dh_params";
       crypto_set_tls_dh_prime(fname);
-      tor_free(fname);
     }
   } else { /* clients don't need a dynamic DH prime. */
     crypto_set_tls_dh_prime(NULL);
diff -uNr tor-0.2.3.15-alpha/src/or/router.c tor-0.2.3.15-alpha-persistent_keys_in_etc/src/or/router.c
--- tor-0.2.3.15-alpha/src/or/router.c	2012-04-29 17:07:09.000000000 -0400
+++ tor-0.2.3.15-alpha-persistent_keys_in_etc/src/or/router.c	2012-05-21 13:05:10.336574459 -0400
@@ -576,10 +576,9 @@
   }
 
   /* 1b. Read identity key. Make it if none is found. */
-  keydir = get_datadir_fname2("keys", "secret_id_key");
+  keydir = "/etc/tor/var/secret_id_key";
   log_info(LD_GENERAL,"Reading/making identity key \"%s\"...",keydir);
   prkey = init_key_from_file(keydir, 1, LOG_ERR);
-  tor_free(keydir);
   if (!prkey) return -1;
   set_server_identity_key(prkey);
 
